### PR TITLE
Add autoignition combustion example.

### PR DESCRIPTION
**Pull request by MTCam**

### CODE REVIEW

File: examples/autoignition-mpi.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+    get_timestep = partial(inviscid_sim_timestep, discr=discr, t=current_t,
+                           dt=current_dt, cfl=current_cfl, eos=eos,
+                           t_final=t_final, constant_cfl=constant_cfl)
+
+    def my_rhs(t, state):
+        cv = split_conserved(dim=dim, q=state)
+        return (inviscid_operator(discr, q=state, t=t,
+                                  boundaries=boundaries, eos=eos)
+                + eos.get_species_source_terms(cv))
```
Comment (MTCam) [inline examples/autoignition-mpi.py:186]: Chemistry source terms added here.

File: examples/autoignition-mpi.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    def my_checkpoint(step, t, dt, state):
+        global checkpoint_t
+        cv = split_conserved(dim, state)
+        reaction_rates = eos.get_production_rates(cv)
+        viz_fields = [("reaction_rates", reaction_rates)]
+        return sim_checkpoint(discr, visualizer, eos, q=state,
+                              vizname=casename, step=step,
+                              t=t, dt=dt, nstatus=nstatus, nviz=nviz,
+                              constant_cfl=constant_cfl, comm=comm,
+                              viz_fields=viz_fields)
```
Comment (MTCam) [inline examples/autoignition-mpi.py:197]: Uses viz_field customization to add species chem reaction rates to viz dumps.

File: examples/autoignition-mpi.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+                           t_final=t_final, constant_cfl=constant_cfl)
+
+    def my_rhs(t, state):
+        cv = split_conserved(dim=dim, q=state)
+        return (inviscid_operator(discr, q=state, t=t,
+                                  boundaries=boundaries, eos=eos)
+                + eos.get_species_source_terms(cv))
+
+    def my_checkpoint(step, t, dt, state):
+        global checkpoint_t
```
Comment (inducer) [inline examples/autoignition-mpi.py:189]: I don't think this works as you intend. Read up on `nonlocal` or boxed values. Maybe discuss at next meeting.

File: examples/autoignition-mpi.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    can_eq_p = cantera_soln.P
+    #    can_eq_e = cantera_soln.int_energy_mass
+    #    can_eq_k = cantera_soln.forward_rate_constants
+    #    can_eq_c = cantera_soln.concentrations
+    #    can_eq_r = cantera_soln.net_rates_of_progress
+    #    can_eq_omega = cantera_soln.net_production_rates
+    if rank == 0:
+        logger.info(init_message)
+        logger.info(f"Expected (p,T,rho, y) = ({can_eq_p}, {can_eq_t},"
+                    f" {can_eq_rho}, {can_eq_y})")
```
Comment (inducer) [inline examples/autoignition-mpi.py:176]: Looks like you're expecting something specific to happen. Could that be `assert`ed? Could you explain? What does `equilibrate` do?

Comment (inducer) [inline examples/autoignition-mpi.py:176]: Read up on `f"{var_name=}"` to avoid having to print labels.

File: examples/autoignition-mpi.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    x[i_fu] = (ox_di_ratio*equiv_ratio)/(stoich_ratio+ox_di_ratio*equiv_ratio)
+    x[i_ox] = stoich_ratio*x[i_fu]/equiv_ratio
+    x[i_di] = (1.0-ox_di_ratio)*x[i_ox]/ox_di_ratio
+    # one_atm = cantera.one_atm
+    one_atm = 101325.0
+
+    print(f"Input state (T,P,X) = ({init_temperature}, {one_atm}, {x}")
+    cantera_soln.TPX = init_temperature, one_atm, x
+    can_t, can_rho, can_y = cantera_soln.TDY
+    can_p = cantera_soln.P
```
Comment (inducer) [inline examples/autoignition-mpi.py:129]: This cantera stuff is mumbo jumbo to me. Could you add some comments to explain?

File: examples/autoignition-mpi.py (hunk 3 above)
Comment (MTCam) [reply examples/autoignition-mpi.py:189]: No, it certainly doesn't.

File: examples/autoignition-mpi.py (hunk 5 above)
Comment (MTCam) [reply examples/autoignition-mpi.py]: Does this help? (bcafb6e)

File: examples/autoignition-mpi.py (hunk 4 above)
Comment (MTCam) [reply examples/autoignition-mpi.py]: What if I want the labels, though?

Comment (MTCam) [reply examples/autoignition-mpi.py]: I added a comment there. Is this enough? I could add Cantera manual links if we want them.
(f2bae70)

File: examples/autoignition-mpi.py (hunk 3 above)
Comment (MTCam) [reply examples/autoignition-mpi.py]: Yeah, we can talk about it.  I have fixed this, but I think the intent is misunderstood in-general.  The issue is that if the simulation stops on a step that was not an output step, then we need to output in order to capture the final state of the sim.  Any one of several situations can cause stepping to exit on a non-output step.  Writing the final state is necessary for restarting from where you stopped, and/or debugging the flow.

Comment (MTCam) [reply examples/autoignition-mpi.py]: Fixed in (f2bae70)

File: examples/autoignition-mpi.py (hunk 4 above)
Comment (MTCam) [reply examples/autoignition-mpi.py]: I think this is what you are looking for here, but I am not sure that this is better than the original message. What do you think? (9db65b7)

Comment (MTCam) [reply examples/autoignition-mpi.py]: I think I've added the required comments explaining the equilibrium situation.  Can you please see if it is more clear now?  (1448345)

**Comment by MTCam** [comment]
Any time you're ready @inducer.  Just wanted to let you know I've finished (with a first stab at) addressing your concerns.

**Review by inducer — APPROVED** [review]
Found just a few nitpicky things. LGTM!

File: examples/autoignition-mpi.py (hunk 4 above)
Comment (inducer) [reply examples/autoignition-mpi.py]: LGTM, but feel free to revert if you prefer the previous style.

File: examples/autoignition-mpi.py (hunk 5 above)
Comment (inducer) [reply examples/autoignition-mpi.py]: Yup, thanks!

File: examples/autoignition-mpi.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+    i_fu = cantera_soln.species_index("C2H4")
+    i_ox = cantera_soln.species_index("O2")
+    i_di = cantera_soln.species_index("N2")
+    x = np.zeros(nspecies)
+    # Set the species mole fractions according to our desired fuel/air mixture
+    x[i_fu] = (ox_di_ratio*equiv_ratio)/(stoich_ratio+ox_di_ratio*equiv_ratio)
+    x[i_ox] = stoich_ratio*x[i_fu]/equiv_ratio
+    x[i_di] = (1.0-ox_di_ratio)*x[i_ox]/ox_di_ratio
+    # one_atm = cantera.one_atm
+    one_atm = 101325.0
```
[the remark this answers is not in the document]
Comment (MTCam) [reply examples/autoignition-mpi.py:134]: Pylint complains that `cantera.one_atm` is undefined, even though it appears to actually work just fine.   I would prefer to use the _Cantera_ version, if I can.

File: examples/autoignition-mpi.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+
+    # Use Cantera for initialization
+    # -- Pick up a CTI for the thermochemistry config
+    from mirgecom.mechanisms import get_mechanism_cti
+    mech_cti = get_mechanism_cti("uiuc")
+
+    cantera_soln = cantera.Solution(phase_id="gas", source=mech_cti)
+    nspecies = cantera_soln.n_species
+
+    # ====== Set up the initial state ========
```
[the remark this answers is not in the document]
Comment (MTCam) [reply examples/autoignition-mpi.py:116]: Ick.  Emacs doesn't do a good job at it.

[the remark this answers is not in the document]
Comment (MTCam) [reply examples/autoignition-mpi.py]: (604b596)

File: examples/autoignition-mpi.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+
+    # Inspection at physics debugging time
+    #    cv = split_conserved(dim, current_state)
+    #    print(f"Initial CV rho: {cv.mass}")
+    #    print(f"Initial CV rhoE: {cv.energy}")
+    #    print(f"Initial CV rhoV: {cv.momentum}")
+    #    print(f"Initial CV rhoY: {cv.mass_fractions}")
+    #    print(f"Initial Y: {cv.mass_fractions / cv.mass}")
+    #    print(f"Initial DV pressure: {eos.pressure(cv)}")
+    #    print(f"Initial DV temperature: {eos.temperature(cv)}")
```
[the remark this answers is not in the document]
Comment (MTCam) [reply examples/autoignition-mpi.py]: (604b596)

File: examples/autoignition-mpi.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+                              viz_fields=viz_fields)
+
+    try:
+        (current_step, current_t, current_state) = \
+            advance_state(rhs=my_rhs, timestepper=timestepper,
+                          checkpoint=my_checkpoint,
+                          get_timestep=get_timestep, state=current_state,
+                          t=current_t, t_final=t_final)
+    except ExactSolutionMismatch as ex:
+        error_state = 1
```
[the remark this answers is not in the document]
Comment (MTCam) [reply examples/autoignition-mpi.py]: (604b596)

File: examples/autoignition-mpi.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MTCam) [reply examples/autoignition-mpi.py]: It "works" when you run it, but `pylint` hates it.  This has been merged, so I'll add a comment and revert to the hard-coded one_atm over on #236 until this weird issue is resolved.
```
************* Module examples.autoignition-mpi
examples/autoignition-mpi.py:135:14: E1101: Module 'cantera' has no 'one_atm' member (no-member)
```

[the remark this answers is not in the document]
Comment (inducer) [reply examples/autoignition-mpi.py]: If pylint is wrong (like it seems to be in this case), you can always shut it up on that line via `# pylint: disable=no-member`.

[the remark this answers is not in the document]
Comment (MTCam) [reply examples/autoignition-mpi.py]: > If pylint is wrong (like it seems to be in this case), you can always shut it up on that line via `# pylint: disable=no-member`.

That's great!  TIL.   (9877cd5).   CI seems wedged, but if it errors out still I'll fix it over on #236.

**Pull request merged**